src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ValidationHelpers.cs

```diff
@@ -57,49 +57,66 @@ public static string GetModelErrorMessageOrDefault(ModelError modelError)
                 {
                     return new[] { ms };
                 }
-
-                return Enumerable.Empty<ModelStateEntry>();
             }
-            else
+            else if (viewData.ModelState.Count > 0)
             {
                 var metadata = viewData.ModelMetadata;
-                var orderer = new ErrorsOrderer(metadata);
+                var modelStateDictionary = viewData.ModelState;
+                var entries = new List<ModelStateEntry>();
+                Visit(modelStateDictionary, modelStateDictionary.Root, metadata, entries);
+
+                Debug.Assert(entries[entries.Count - 1] == modelStateDictionary.Root);
+                // Special case the root node. It is only 'Visible` if it has any errors associated with it.
+                if (modelStateDictionary.Root.Errors.Count == 0)
+                {
+                    entries.RemoveAt(entries.Count - 1);
+                }
+
+                if (entries.Count != modelStateDictionary.Count)
+                {
+                    // Account for entries in the ModelStateDictionary that do not have corresponding ModelMetadata values.
+                    foreach (var entry in modelStateDictionary)
+                    {
+                        if (!entries.Contains(entry.Value))
+                        {
+                            entries.Add(entry.Value);
+                        }
+                    }
+                }
 
-                return viewData.ModelState
-                    .OrderBy(data => orderer.GetOrder(data.Key))
-                    .Select(ms => ms.Value);
+                return entries;
             }
+
+            return EmptyModelStateEntries;
         }
 
-        // Helper for sorting modelStates to respect the ordering in the metadata.
-        // ModelState doesn't refer to ModelMetadata, but we can correlate via the property name.
-        private class ErrorsOrderer
+        private static void Visit(
```

Comment:
This depth-first then `Properties`-based ordering is unfortunately incorrect. The end-to-end intent is to sort the errors to match the order `@Html.EditorFor()` would generate. Previous code hit that nail on the head, albeit accidentally -- did a depth-first, then `Properties`, then `ICollection<T>`-based ordering. Everything about collections is missing here. Should be restored.